### PR TITLE
Merging tidegauge and tidegaugeMultiple

### PR DESCRIPTION
**Pull request by davbyr**
I've started the process of trying to merge these two objects. What it boils down to is making it so all the routines in the original Tidegauge object are expecting a dataset with an additional 'id' dimension. So even if there is only one location, there is still an 'id' dimension. Here's a list of jobs:

> Update GESLA reading routines to have an id dimension
> Update BODC reading routines to output id dimension
> Update HLW reading routines to output id dimension
> GESLA read accepts multiple files or glob and returns multiple datasets
[ NO ] BODC read accepts multiple files or glob and returns multiple datasets
[ NO ] HLW read accepts multiple files or glob and returns multiple datasets
> Remove methods that create 'multiple' tidegauge objects
> Update analysis methods to work with 'id' dimension
> Moved across analysis methods from TidegaugeMultiple to TidegaugeAnalysis
> All reading routines should now make site_name or similar as a variable not an attribute. 
> Updated docstrings
> Unit test updated
> Correct approach to creating a tidegauge object is no longer default GESLA.

Also MORE updates to unit testing system:

There are now two settings booleans at the top of unit_test.py: `generate_unit_test_contents` and `calculate_coverage`. For the former, I have moved the generate_unit_test_contents scripts into unit_test.py. Enable this to write a new unit_test_contents scripts based on the tests in tests_to_do.

`calculate_coverage` used the `coverage` library to estimate the % coverage of the current unit tests. This tells you how much of each COAsT file is 'touched' by the unit test as well as a % for the entire package. Really useful I think. This is optional and requires a separate library to be installed called `coverage`. This can be done using `pip install coverage`. The unit test will only import if `calculate_coverage = True`.

Coverage report looks like:

```
Name                                           Stmts   Miss  Cover
------------------------------------------------------------------
coast/altimetry.py                               169     46    73%
coast/climatology.py                              85     71    16%
coast/coast.py                                   205    142    31%
...
```

### CODE REVIEW

**Comment by jpolton** [comment]
Had a quick skim:

* Is Tidegauge[time] --> Tidegauge[id,time] a breaking change?

* "sea_level" --> "ssh" seems similarly specific though more ambiguous. If changing the variable name then "height" or "level" would be more generic?

**Comment by jpolton** [comment]
FYI You might be interested to know that there is a new GESLA format (v5) E.g. collated by your favourite external examiner

```
# FORMAT VERSION 5.0 Web: https://gesla.org Email: gesla.help@gmail.com
# SITE NAME Southampton
# SITE CODE SOU
# COUNTRY GBR
# CONTRIBUTOR Ivan D. Haigh
# CONTRIBUTOR WEBSITE https://ivanhaigh.com
# CONTRIBUTOR CONTACT I.D.Haigh@soton.ac.uk
# ORGINATOR Ivan D. Haigh
# ORIGINATOR WEBSITE https://ivanhaigh.com
# ORIGINATOR CONTACT I.D.Haigh@soton.ac.uk
# LATITUDE      50.88390000
# LONGITUDE     -1.39310000
# COORDINATE SYSTEM Unspecified
# START DATE/TIME 1935/01/01 00:00:00
# END DATE/TIME 1990/12/31 23:45:00
# NUMBER OF YEARS 54
# TIME ZONE HOURS 0
# DATUM INFORMATION Admiralty Chart Datum (ACD) 
# INSTRUMENT Unspecified
# PRECISION Unspecified
# NULL VALUE -99.9999
# GAUGE TYPE Coastal
# OVERALL RECORD QUALITY No obvious issues
# 
# CREATION DATE UTC 2021/11/03
# 
# COLUMN 1 Date yyyy/mm/dd
# COLUMN 2 Time hh:mm:ss
# COLUMN 3 Observed sea level (m)
# COLUMN 4 Observed sea level QC flag
# COLUMN 5 Use-in-analysis flag (1 = use, 0 = do not use)
# 
# Quality-control (QC) flags for column 4
# 
# 0 - no quality control
# 1 - correct value 
# 2 - interpolated value
# 3 - doubtful value
# 4 - isolated spike or wrong value
# 5 - missing value
# 
1935/01/01 00:00:00 -99.9999 5 0
```
snip snip
```
1935/01/05 09:30:00   3.7705 1 1
1935/01/05 09:45:00   4.0634 1 1
1935/01/05 10:00:00   4.2748 1 1
1935/01/05 10:15:00   4.4337 1 1
1935/01/05 10:30:00   4.5107 1 1
1935/01/05 10:45:00   4.5286 1 1
1935/01/05 11:00:00   4.4973 1 1
1935/01/05 11:15:00   4.4440 1 1
1935/01/05 11:30:00   4.4215 1 1
1935/01/05 11:45:00   4.4240 1 1
1935/01/05 12:00:00   4.4636 1 1
1935/01/05 12:15:00   4.5194 1 1
1935/01/05 12:30:00   4.5389 1 1
1935/01/05 12:45:00   4.5199 1 1
1935/01/05 13:00:00   4.4448 1 1
1935/01/05 13:15:00   4.3217 1 1
1935/01/05 13:30:00   4.2249 1 1
1935/01/05 13:45:00   4.0532 1 1
1935/01/05 14:00:00   3.8163 1 1
1935/01/05 14:15:00   3.5733 1 1
1935/01/05 14:30:00   3.2116 1 1
1935/01/05 14:45:00   2.7753 1 1
1935/01/05 15:00:00   2.2854 1 1
1935/01/05 15:15:00   1.7838 1 1
1935/01/05 15:30:00   1.3097 1 1
1935/01/05 15:45:00   0.9026 1 1
1935/01/05 16:00:00   0.6002 1 1
1935/01/05 16:15:00   0.4311 1 1
1935/01/05 16:30:00   0.4082 1 1
1935/01/05 16:45:00   0.4680 1 1
1935/01/05 17:00:00   0.6325 1 1
1935/01/05 17:15:00   0.8729 1 1
1935/01/05 17:30:00   1.0763 1 1
1935/01/05 17:45:00   1.2740 1 1
1935/01/05 18:00:00   1.4393 1 1
1935/01/05 18:15:00   1.6384 1 1
1935/01/05 18:30:00   1.7753 1 1
1935/01/05 18:45:00   1.8812 1 1
1935/01/05 19:00:00   1.9826 1 1
1935/01/05 19:15:00   2.0862 1 1
1935/01/05 19:30:00   2.1560 1 1
1935/01/05 19:45:00   2.1939 1 1
1935/01/05 20:00:00   2.2001 1 1
1935/01/05 20:15:00   2.2098 1 1
1935/01/05 20:30:00   2.2490 1 1
1935/01/05 20:45:00   2.3874 1 1
1935/01/05 21:00:00   2.7411 1 1
1935/01/05 21:15:00   2.9942 1 1
1935/01/05 21:30:00   3.1934 1 1
1935/01/05 21:45:00   3.4067 1 1
1935/01/05 22:00:00   3.6663 1 1
1935/01/05 22:15:00   3.9273 1 1
1935/01/05 22:30:00   4.1905 1 1
1935/01/05 22:45:00   4.4111 1 1
1935/01/05 23:00:00   4.4662 1 1
1935/01/05 23:15:00   4.4502 1 1
1935/01/05 23:30:00   4.3825 1 1
1935/01/05 23:45:00   4.3349 1 1
1935/01/06 00:00:00   4.3286 1 1
1935/01/06 00:15:00   4.3771 1 1
1935/01/06 00:30:00   4.4682 1 1
1935/01/06 00:45:00   4.5525 1 1
1935/01/06 01:00:00   4.5888 1 1
1935/01/06 01:15:00   4.5706 1 1
1935/01/06 01:30:00   4.4929 1 1
1935/01/06 01:45:00   4.3750 1 1
1935/01/06 02:00:00   4.2285 1 1
1935/01/06 02:15:00   4.0664 1 1
1935/01/06 02:30:00   3.8369 1 1
1935/01/06 02:45:00   3.5417 1 1
1935/01/06 03:00:00   3.1661 1 1
1935/01/06 03:15:00   2.7313 1 1
1935/01/06 03:30:00   2.2672 1 1
1935/01/06 03:45:00   1.7634 1 1
1935/01/06 04:00:00   1.2900 1 1
1935/01/06 04:15:00   0.9082 1 1
1935/01/06 04:30:00   0.6283 1 1
1935/01/06 04:45:00   0.4920 1 1
1935/01/06 05:00:00   0.5188 1 1
1935/01/06 05:15:00   0.6392 1 1
1935/01/06 05:30:00   0.8147 1 1
1935/01/06 05:45:00   1.0517 1 1
1935/01/06 06:00:00   1.2777 1 1
```

**Comment by davbyr** [comment]
@jpolton Yes I think multiple of the changes in this branch are breaking changes seeing as I have had to change the unit tests too.

**Comment by anwiseNOCL** [comment]
@davbyr I've just being running this and it threw up a few errors. I think the weird error you were getting that go away when you apply np.copy are being caused by the fact that you've changed the chunk size in the t_grid.json file to 10 but there are only 7 time steps. This is probably resulting in some weird memory behaviour that is causing those errors. At least I found that reducing it back to 2 got rid of one in the **_coast.Gridded.get_e3_from_ssh_** method. Currently re running unit test to see if it got rid of the rest of the fails.

**Comment by davbyr** [comment]
@anwiseNOCL huh that's interesting, well spotted Ant. I don't remember doing that at all but will switch it back.

**Comment by anwiseNOCL** [comment]
Getting these issue:
```
======================================================================
ERROR: test_find_maxima (test_stats_utilities.test_stats_utilities)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/users/anwise/COAsT_v1/unit_testing/test_stats_utilities.py", line 18, in test_find_maxima
    tg = coast.Tidegauge(files.fn_tidegauge, date_start=date0, date_end=date1)
TypeError: __init__() got an unexpected keyword argument 'date_start'

======================================================================
ERROR: test_tidegauge_tutorial (test_example_scripts.test_example_scripts)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/users/anwise/COAsT_v1/unit_testing/test_example_scripts.py", line 18, in test_tidegauge_tutorial
    from example_scripts import tidegauge_tutorial
  File "/home/users/anwise/COAsT_v1/example_scripts/tidegauge_tutorial.py", line 33, in <module>
    tidegauge = coast.Tidegauge(file_path=fn_tidegauge, date_start=date_0, date_end=date_1, config=fn_tidegauge_config)
TypeError: __init__() got an unexpected keyword argument 'file_path'

======================================================================
ERROR: test_tidetable_tutorial (test_example_scripts.test_example_scripts)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/users/anwise/COAsT_v1/unit_testing/test_example_scripts.py", line 23, in test_tidetable_tutorial
    from example_scripts import tidetable_tutorial
  File "/home/users/anwise/COAsT_v1/example_scripts/tidetable_tutorial.py", line 40, in <module>
    tg.dataset = tg.read_hlw_to_xarray(filnam, date_start, date_end)
AttributeError: 'Tidegauge' object has no attribute 'read_hlw_to_xarray'

----------------------------------------------------------------------
Ran 66 tests in 322.752s

FAILED (errors=3)
```

**Comment by davbyr** [comment]
@anwiseNOCL I have strange errors popping up where previously I'm sure the unit test was fine....... So I may need to repush anyway.

**Comment by davbyr** [comment]
@anwiseNOCL unit test should work now.

Also pushed some more changes to unit_test.py. See description. Just some little geeky things

File: coast/gridded.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
 from .coast import Coast
 from .config_parser import ConfigParser
 from .logging_util import get_slug, debug, info, warn, error, warning
+from . import logging_util
```
Comment (jpolton) [inline coast/gridded.py:14]: looks a bit repetative with `logging_util`?

File: .gitignore [hunk 2]
```diff
@@ -26,4 +26,5 @@ unit_testing/*log
 sandbox.py
 
 config/nemo_tst_gridded.json
-
+.coverage
```
Comment (jpolton) [inline .gitignore:29]: This will need some explaining somewhere. Perhaps a ticket to add some docs

File: coast/tidegauge.py [hunk 3]
```diff
@@ -888,68 +845,14 @@ def read_bodc_data(fn_bodc, date_start=None, date_end=None, header_length: int =
...
-        ax.set_ylim((min(Y) - 10, max(Y) + 10))
-        return fig, ax
-
-    def plot_timeseries(self, var_list=["sea_level"], date_start=None, date_end=None, plot_line=False):
+    def plot_timeseries(self, id, var_list=["ssh"], date_start=None, date_end=None, plot_line=False):
         """
         Quick plot of time series stored within object's dataset
         Parameters
         ----------
         date_start (datetime) : Start date for plotting
```
Comment (jpolton) [inline coast/tidegauge.py:853]: add "id"

**Review by jpolton — COMMENTED** [review]
tidegauge_validation_tutorial.py still has tidegaugeMultiple in it. But this doesn't throw up an error when I unit test locally ....

**Comment by jpolton** [comment]
Fundamentally this passes the unit testing.
Look through the changes there are A LOT of things done within Tidegauge() that will take me some time to ingest. For example separating out analysis and loading methods. This seems sensible.

**Comment by jpolton** [comment]
I spotted xarray.ufuncs: 
Screenshot 2022-03-31 at 21 23 11

I think that these should just be `numpy` methods now with the numpy refresh

**Comment by davbyr** [comment]
Thanks @jpolton 

The example scripts work OK locally for me. Anna was having very similar issues with the example scripts this evening.

I'm confused as to why xarray ufuncs is popping up in files such as contour and profile when I haven't changed them in this branch. Tidegauge_multiple doesnt matter as it is now redundant. I will remove it from the repo

File: .gitignore (hunk 2 above)
Comment (davbyr) [reply .gitignore:29]: Docs already added to develop branch of site

**Review by anwiseNOCL — APPROVED** [review]
Unit test works for me on jasmin without issue. I like that coverage addition to the unit test as well. It's a good idea to keep the coverage high where possible. Only thing is it broke the test
when the example scripts are being tested as well. If i comment those out i get:

----------------------------------------------------------------------
Ran 58 tests in 283.181s

OK
 
COVERAGE REPORT (Classes): 
Name                          Stmts   Miss  Cover
-------------------------------------------------
coast/altimetry.py              167     44    74%
coast/climatology.py             85     71    16%
coast/coast.py                  205    142    31%
coast/config_parser.py           51     16    69%
coast/contour.py                485     84    83%
coast/crps_util.py              117     43    63%
coast/eof.py                     83      8    90%
coast/general_utils.py          153     84    45%
coast/gridded.py                377     93    75%
coast/index.py                   49     21    57%
coast/internal_tide.py           81     10    88%
coast/logging_util.py            50     32    36%
coast/mask_maker.py              66     36    45%
coast/plot_util.py              129     32    75%
coast/profile.py                341     92    73%
coast/stats_util.py              67     18    73%
coast/tidegauge.py              455    154    66%
coast/tidegauge_analysis.py     189     52    72%
coast/transect.py               420     75    82%
coast/xesmf_convert.py           32     17    47%
-------------------------------------------------
TOTAL                          3602   1124    69%

**Review by jpolton — APPROVED** [review]

**Pull request merged**